### ISSUE TITLE
How do Transformers Work in NLP? A Guide to the Latest State-of-the-Art Models

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10514]
How do Transformers Work in NLP? A Guide to the Latest State-of-the-Art Models<br>
<br>
https://ift.tt/L9kqjUs<br>
<br>
Prateek Joshi<br>
<br>

<div><div><div><div><section><h2 class="wp-block-heading">Overview</h2>
<ul>
<li>The Transformer model in NLP has truly changed the way we work with text data</li>
<li>Transformer is behind the recent NLP developments, including Google’s BERT</li>
<li>Learn how the Transformer idea works, how it’s related to language modeling, sequence-to-sequence modeling, and how it enables Google’s BERT model</li>
</ul>
<h2 class="wp-block-heading">Introduction</h2>
<p>I love being a data scientist working in <a rel="noopener noreferrer" href="https://courses.analyticsvidhya.com/courses/natural-language-processing-nlp?utm_source=blog&amp;utm_medium=understanding-transformers-nlp-state-of-the-art-models" target="_blank">Natural Language Processing (NLP)</a> and learning through <a href="https://blackbelt.analyticsvidhya.com/bundles/natural-language-processing-course-online-nlp-training">NLP Training</a> right now. The breakthroughs and developments are occurring at an unprecedented pace. From the super-efficient ULMFiT framework to Google’s BERT, NLP is truly in the midst of a golden era.</p>
<p>And at the heart of this revolution is the concept of the Transformer. This has transformed the way we data scientists work with text data – and you’ll soon see how in this article.</p>
<p>Want an example of how useful Transformer is? Take a look at the paragraph below:</p>
<p>The highlighted words refer to the same person – Griezmann, a popular football player. It’s not that difficult for us to figure out the relationships among such words spread across the text. However, it is quite an uphill task for a machine.</p>
<p>Capturing such relationships and sequences of words in sentences is vital for a machine to understand a natural language. <strong>This is where the Transformer concept plays a major role.</strong></p>
<p><em>Note: This article assumes a basic understanding of a few deep learning concepts:</em></p>
<h2 class="wp-block-heading">Sequence-to-Sequence Models – A Backdrop</h2>
<p><strong>Sequence-to-sequence (seq2seq)</strong> models in NLP are used to convert sequences of Type A to sequences of Type B. For example, translation of English sentences to German sentences is a sequence-to-sequence task.</p>
<p><b>Recurrent Neural Network (RNN) based sequence-to-sequence models</b> have garnered a lot of traction ever since they were introduced in 2014. Most of the data in the current world are in the form of sequences – it can be a number sequence, text sequence, a video frame sequence or an audio sequence.</p>
<p>The performance of these seq2seq models was further enhanced with the addition of the <b>Attention Mechanism </b>in 2015. How quickly advancements in NLP have been happening in the last 5 years – incredible!</p>
<p>These sequence-to-sequence models are pretty versatile and they are used in a variety of NLP tasks, such as:</p>
<ul>
<li>Machine Translation</li>
<li>Text Summarization</li>
<li>Speech Recognition</li>
<li>Question-Answering System, and so on</li>
</ul>
<h3 class="wp-block-heading">RNN based Sequence-to-Sequence Model</h3>
<p>Let’s take a simple example of a sequence-to-sequence model. Check out the below illustration:</p>
<div class="wp-block-image">
<figure><a href="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/seq2seq.gif"><img alt="sequence to sequence, nlp" src="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/seq2seq.gif" class="wp-image-53905"></a><figcaption class="wp-element-caption"><em>German to English Translation using seq2seq</em></figcaption></figure>
</div>
<p>The above seq2seq model is converting a German phrase to its English counterpart. Let’s break it down:</p>
<ul>
<li>Both <b>Encoder</b> and <b>Decoder</b> are RNNs</li>
<li>At every time step in the Encoder, the RNN takes a word vector (xi) from the input sequence and a hidden state (Hi) from the previous time step</li>
<li>The hidden state is updated at each time step</li>
<li>The hidden state from the last unit is known as the <b>context vector.</b> This contains information about the input sequence</li>
<li>This context vector is then passed to the decoder and it is then used to generate the target sequence (English phrase)</li>
<li>If we use the <b>Attention mechanism</b>, then the weighted sum of the hidden states are passed as the context vector to the decoder</li>
</ul>
<h3 class="wp-block-heading">Challenges</h3>
<p>Despite being so good at what it does, there are certain limitations of seq-2-seq models with attention:</p>
<ul>
<li>Dealing with long-range dependencies is still challenging</li>
<li>The sequential nature of the model architecture prevents parallelization. These challenges are addressed by Google Brain’s Transformer concept</li>
</ul>
<h2 class="wp-block-heading"><b>Introduction to the Transformer</b></h2>
<p>The Transformer in NLP is a novel architecture that aims to solve sequence-to-sequence tasks while handling long-range dependencies with ease. The Transformer was proposed in the paper <a rel="nofollow noopener noreferrer" href="https://arxiv.org/abs/1706.03762" target="_blank">Attention Is All You Need</a>. It is recommended reading for anyone interested in NLP.</p>
<p>Quoting from the paper:</p>
<blockquote class="wp-block-quote">
<p>“The Transformer is the first transduction model relying entirely on self-attention to compute representations of its input and output without using sequence-aligned RNNs or convolution.”</p>
</blockquote>
<p>Here, “transduction” means the conversion of input sequences into output sequences. The idea behind Transformer is to handle the dependencies between input and output with <b>attention </b>and recurrence completely.</p>
<p>Let’s take a look at the architecture of the Transformer below. It might look intimidating but don’t worry, we will break it down and understand it block by block.</p>
<h3 class="wp-block-heading">Understanding Transformer’s Model Architecture</h3>
<div class="wp-block-image">
<figure><a href="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-17-19-53-10.png"><img alt="transformer" src="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-17-19-53-10.png" class="wp-image-54015"></a><figcaption class="wp-element-caption">The Transformer – Model Architecture<br>(Source: https://ift.tt/JFnlm7f)</figcaption></figure>
</div>
<p>The above image is a superb illustration of Transformer’s architecture. Let’s first focus on the <b>Encoder</b> and <b>Decoder</b> parts only.</p>
<p>Now focus on the below image. The Encoder block has 1 layer of a <b>Multi-Head Attention</b> followed by another layer of <b>Feed Forward Neural Network</b>. The decoder, on the other hand, has an extra <b>Masked</b> <b>Multi-Head Attention.</b></p>
<p><strong>The encoder and decoder blocks are actually multiple identical encoders and decoders stacked on top of each other.</strong> Both the encoder stack and the decoder stack have the same number of units.</p>
<p>The number of encoder and decoder units is a hyperparameter. In the paper, 6 encoders and decoders have been used.</p>
<p>Let’s see how this setup of the encoder and the decoder stack works:</p>
<ul>
<li>The word embeddings of the input sequence are passed to the first encoder</li>
<li>These are then transformed and propagated to the next encoder</li>
<li>The output from the last encoder in the encoder-stack is passed to all the decoders in the decoder-stack as shown in the figure below:</li>
</ul>
<p>An important thing to note here – in addition to the <strong>self-attention</strong> and feed-forward layers, the decoders also have one more layer of Encoder-Decoder Attention layer. This helps the decoder focus on the appropriate parts of the input sequence.</p>
<p>You might be thinking – what exactly does this “Self-Attention” layer do in the Transformer? Excellent question! This is arguably the most crucial component in the entire setup so let’s understand this concept.</p>
<h3 class="wp-block-heading">Getting a Hang of Self-Attention</h3>
<p>According to the paper:</p>
<blockquote class="wp-block-quote">
<p>“Self-attention, sometimes called intra-attention, is an attention mechanism relating different positions of a single sequence in order to compute a representation of the sequence.”</p>
</blockquote>
<p>Take a look at the above image. Can you figure out what the term <b>“it”</b> in this sentence refers to?</p>
<p>Is it referring to the street or to the animal? It’s a simple question for us but not for an algorithm. When the model is processing the word “it”, self-attention tries to associate “it” with <b>“animal”</b> in the same sentence.</p>
<p>Self-attention allows the model to look at the other words in the input sequence to get a better understanding of a certain word in the sequence. Now, let’s see how we can calculate self-attention.</p>
<h3 class="wp-block-heading">Calculating Self-Attention</h3>
<p>I have divided this section into various steps for ease of understanding.</p>
<p>First, we need to create three vectors from each of the encoder’s input vectors:</p>
<ol>
<li>Query Vector</li>
<li>Key Vector</li>
<li>Value Vector.</li>
</ol>
<p>These vectors are trained and updated during the training process. We’ll know more about their roles once we are done with this section</p>
<p>Next, we will calculate self-attention for every word in the input sequence</p>
<p>Consider this phrase – “Action gets results”. To calculate the self-attention for the first word “Action”, we will calculate scores for all the words in the phrase with respect to “Action”. This score determines the importance of other words when we are encoding a certain word in an input sequence</p>
<p>1. The score for the first word is calculated by taking the dot product of the Query vector (q1) with the keys vectors (k1, k2, k3) of all the words:</p>
<p>2. Then, these scores are divided by 8 which is the square root of the dimension of the key vector:</p>
<p>3. Next, these scores are normalized using the softmax activation function:</p>
<p>4. These normalized scores are then multiplied by the value vectors (v1, v2, v3) and sum up the resultant vectors to arrive at the final vector (z1). This is the output of the self-attention layer. It is then passed on to the feed-forward network as input.</p>
<p>So, z1 is the self-attention vector for the first word of the input sequence “Action gets results”. We can get the vectors for the rest of the words in the input sequence in the same fashion:</p>
<p>Self-attention is computed not once but multiple times in the Transformer’s architecture, in parallel and independently. It is therefore referred to as <strong>Multi-head Attention</strong>. The outputs are concatenated and linearly transformed as shown in the figure below:</p>
<p>According to the paper “Attention Is All You Need”:</p>
<blockquote class="wp-block-quote">
<p>“Multi-head attention allows the model to jointly attend to information from different representation subspaces at different positions.”</p>
</blockquote>
<p>You can access the code to implement Transformer <a rel="nofollow noopener noreferrer" href="https://paperswithcode.com/paper/attention-is-all-you-need" target="_blank">here</a>.</p>
<h3 class="wp-block-heading">Limitations of the Transformer</h3>
<p>Transformer is undoubtedly a huge improvement over the RNN based seq2seq models. But it comes with its own share of limitations:</p>
<ul>
<li>Attention can only deal with fixed-length text strings. The text has to be split into a certain number of segments or chunks before being fed into the system as input</li>
<li>This chunking of text causes <b>context fragmentation</b>. For example, if a sentence is split from the middle, then a significant amount of context is lost. In other words, the text is split without respecting the sentence or any other semantic boundary</li>
</ul>
<p>So how do we deal with these pretty major issues? That’s the question folks who worked with Transformer asked. And out of this came Transformer-XL.</p>
<h2 class="wp-block-heading">Understanding Transformer-XL</h2>
<p>Transformer architectures can learn longer-term dependency. However, they can’t stretch beyond a certain level due to the use of fixed-length context (input text segments). A new architecture was proposed to overcome this shortcoming in the paper – <a rel="noopener noreferrer" href="https://arxiv.org/pdf/1901.02860.pdf" target="_blank">Transformer-XL: Attentive Language Models Beyond a Fixed-Length Context</a>.</p>
<p>In this architecture, the hidden states obtained in previous segments are reused as a source of information for the current segment. It enables modeling longer-term dependency as the information can flow from one segment to the next.</p>
<h3 class="wp-block-heading">Using Transformer for Language Modeling<b><br></b><b></b>
</h3>
<blockquote class="wp-block-quote">
<p>Think of language modeling as a process of estimating the probability of the next word given the previous words.</p>
</blockquote>
<p><a rel="noopener noreferrer" href="https://arxiv.org/abs/1808.04444" target="_blank">Al-Rfou et al. (2018)</a> proposed the idea of <b>applying the Transformer model for language modeling</b>. As per the paper, the entire corpus can be split into fixed-length segments of manageable sizes. Then, we train the Transformer model on the segments independently, ignoring all contextual information from previous segments:</p>
<div class="wp-block-image">
<figure><a href="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-18-08-20-09.png"><img alt="transformer nlp" src="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-18-08-20-09.png" class="wp-image-54029"></a><figcaption class="wp-element-caption"><em>Transformer Model with a segment length of 4 (Source: https://ift.tt/CVzXaBZ)</em></figcaption></figure>
</div>
<p>This architecture doesn’t suffer from the problem of vanishing gradients. But the context fragmentation limits its longer-term dependency learning. During the evaluation phase, the segment is shifted to the right by only one position. The new segment has to be processed entirely from scratch. This evaluation method is unfortunately quite compute-intensive.</p>
<h3 class="wp-block-heading">Using Transformer-XL for Language Modeling</h3>
<p>During the training phase in Transformer-XL, the hidden state computed for the previous state is used as an additional context for the current segment. This recurrence mechanism of Transformer-XL takes care of the limitations of using a fixed-length context.</p>
<div class="wp-block-image">
<figure><a href="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-18-08-32-41.png"><img alt="transformer xl" src="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-18-08-32-41.png" class="wp-image-54031"></a><figcaption class="wp-element-caption"><em>Transformer XL Model with a segment length of 4</em></figcaption></figure>
</div>
<p>During the evaluation phase, the representations from the previous segments can be reused instead of being computed from scratch (as is the case of the Transformer model). This, of course, increases the computation speed manifold.</p>
<p>You can access the code to implement Transformer-XL <a rel="noopener noreferrer" href="https://paperswithcode.com/paper/transformer-xl-attentive-language-models" target="_blank">here</a>.</p>
<h2 class="wp-block-heading"><b>The New Sensation in NLP: Google’s BERT (Bidirectional Encoder Representations from Transformers)</b></h2>
<p>We all know how significant <b>transfer learning</b> has been in the field of computer vision. For instance, a pre-trained deep learning model could be fine-tuned for a new task on the ImageNet dataset and still give decent results on a relatively small labeled dataset.</p>
<p>Language model pre-training similarly has been quite effective for improving many natural language processing tasks: (<a rel="noopener noreferrer" href="https://openai.com/blog/language-unsupervised/" target="_blank">https://paperswithcode.com/paper/transformer-xl-attentive-language-models</a> and <a rel="noopener noreferrer" href="https://arxiv.org/abs/1801.06146" target="_blank">https://paperswithcode.com/paper/transformer-xl-attentive-language-models</a>).</p>
<p>The BERT framework, a new language representation model from Google AI, uses pre-training and fine-tuning to create state-of-the-art models for a wide range of tasks. These tasks include question answering systems, sentiment analysis, and language inference.</p>
<h3 class="wp-block-heading">BERT’s Model Architecture</h3>
<p>BERT uses a multi-layer bidirectional Transformer encoder. Its self-attention layer performs self-attention in both directions. Google has released two variants of the model:</p>
<ol>
<li>
<b>BERT Base</b>: Number of Transformers layers = 12, Total Parameters = 110M</li>
<li>
<b>BERT Large</b>: Number of Transformers layers = 24, Total Parameters = 340M</li>
</ol>
<p>BERT uses bidirectionality by pre-training on a couple of tasks — <b>Masked Language Model</b> and <b>Next Sentence Prediction</b>. Let’s discuss these two tasks in detail.</p>
<h3 class="wp-block-heading">BERT Pre-Training Tasks</h3>
<p>BERT is pre-trained using the following two unsupervised prediction tasks.</p>
<h4 class="wp-block-heading">1. Masked Language Modeling (MLM)</h4>
<p>According to the <a rel="nofollow noopener noreferrer" href="https://arxiv.org/pdf/1810.04805.pdf" target="_blank">paper</a>:</p>
<blockquote class="wp-block-quote">
<p>“The masked language model randomly masks some of the tokens from the input, and the objective is to predict the original vocabulary id of the masked word based only on its context. Unlike left-to-right language model pre-training, the MLM objective allows the representation to fuse the left and the right context, which allows us to pre-train a deep bidirectional Transformer.”</p>
</blockquote>
<p>The Google AI researchers masked 15% of the words in each sequence at random. The task? To predict these masked words. A caveat here – the masked words were not always replaced by the masked tokens [MASK] because the [MASK] token would never appear during fine-tuning.</p>
<p>So, the researchers used the below technique:</p>
<ul>
<li>80% of the time the words were replaced with the masked token [MASK]</li>
<li>10% of the time the words were replaced with random words</li>
<li>10% of the time the words were left unchanged</li>
</ul>
<h4 class="wp-block-heading">2. Next Sentence Prediction</h4>
<p>Generally, language models do not capture the relationship between consecutive sentences. BERT was pre-trained on this task as well.</p>
<p>For language model pre-training, BERT uses pairs of sentences as its training data. The selection of sentences for each pair is quite interesting. Let’s try to understand it with the help of an example.</p>
<p>Imagine we have a text dataset of 100,000 sentences and we want to pre-train a BERT language model using this dataset. So, there will be 50,000 training examples or pairs of sentences as the training data.</p>
<ul>
<li>For 50% of the pairs, the second sentence would actually be the next sentence to the first sentence</li>
<li>For the remaining 50% of the pairs, the second sentence would be a random sentence from the corpus</li>
<li>The labels for the first case would be <em><b>‘IsNext’</b></em> and <em><b>‘NotNext’ </b></em>for the second case</li>
</ul>
<p>Architectures like BERT demonstrate that unsupervised learning (pre-training and fine-tuning) is going to be a key element in many language understanding systems. Low resource tasks especially can reap huge benefits from these deep <i>bidirectional</i> architectures.</p>
<p>Below is a snapshot of a few NLP tasks where BERT plays an important role:</p>
<div class="wp-block-image">
<figure><a href="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-18-08-50-57.png"><img alt="BERT use cases" src="https://cdn.analyticsvidhya.com/wp-content/uploads/2019/06/Screenshot-from-2019-06-18-08-50-57.png" class="wp-image-54034"></a><figcaption class="wp-element-caption"><em>Source: https://ift.tt/QXY6AJw</em></figcaption></figure>
</div>
<h2 class="wp-block-heading">End Notes</h2>
<p>We should really consider ourselves lucky as so many state-of-the-art advancements are happening in NLP at such a rapid pace. Architectures like Transformers and BERT are paving the way for even more advanced breakthroughs to happen in the coming years.</p>
<p>I encourage you to implement these models and share your work in the comments section below. And if you have any feedback on this article or any doubts/queries, then do let me know and I will get back to you.</p>
<p>You can also take the below course to learn or brush up your NLP skills:</p>
</section></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Analytics Vidhya <br>
<br>
January 8, 2024 at 11:25AM
